server/src/main/java/org/elasticsearch/common/compress/DeflateCompressor.java

```diff
@@ -73,6 +74,59 @@ public int headerLength() {
         return HEADER.length;
     }
 
+    // Reusable inflater reference for streaming decompression
+    private static final ThreadLocal<ReleasableReference<Inflater>> inflaterForStreamRef = ThreadLocal.withInitial(() -> {
+        final Inflater inflater = new Inflater(true);
+        return new ReleasableReference<>(inflater, inflater::reset);
+    });
+
+    // Reusable deflater reference for streaming compression
+    private static final ThreadLocal<ReleasableReference<Deflater>> deflaterForStreamRef = ThreadLocal.withInitial(() -> {
+        final Deflater deflater = new Deflater(LEVEL, true);
+        return new ReleasableReference<>(deflater, deflater::reset);
+    });
+
+    // Reference to a deflater or inflater that is used to make sure we do not use the same stream twice when nesting streams.
+    private static final class ReleasableReference<T> implements Releasable {
+
+        protected final T resource;
+
+        private final Releasable releasable;
+
+        // Thread that is currently using this reference
+        private Thread user = null;
```

Comment:
s/user/thread ?

Also maybe add a note that this is only used when assertions are enabled?